### PR TITLE
INT-4248: Refactor `RedisLockRegistry`

### PR DESCRIPTION
**Pull request by vpavic**
JIRA: https://jira.spring.io/browse/INT-4248

This is still WIP, please review @artembilan. The changes to `RedisLockRegistry` are quite significant and the implementation is now very similar to one in `JdbcLockRegistry`.

Tests are a mess ATM, I'll address them after you confirm the approach taken here is OK with you. I did however include a test that demonstrate the issue originally described in Jira issue - take a look at `RedisLockRegistryTests#testExpireTwoRegistries`.

### CODE REVIEW

**Comment by artembilan** [comment]
Looks awesome at a glance, @vpavic !

I'm pretty sure you can go ahead with tests polishing.
I'll review properly tomorrow.

Only concern for now that we can't just remove public API in the point version. It must be `@Deprecated` with an appropriate JavaDocs.
Also I don't think that we need to keep that script in the separate file. Unlike SQL DDL scripts, this one isn't for installation in the end-user environment - just for invocation, like `SELECT`s.

Thank you very much!

**Comment by vpavic** [comment]
Thanks @artembilan!

> Only concern for now that we can't just remove public API in the point version. It must be `@Deprecated` with an appropriate JavaDocs.
Also I don't think that we need to keep that script in the separate file. Unlike SQL DDL scripts, this one isn't for installation in the end-user environment - just for invocation, like `SELECT`s.

I've updated the PR to address this. Will revisit the tests later today.

**Comment by artembilan** [comment]
> I've updated the PR to address this.

Please, don't squash commits for PR.
It is much easier to trace changed during by the commit history.
We do squash during merge into `upstream`.

**Comment by vpavic** [comment]
> Please, don't squash commits for PR.
It is much easier to trace changes during review by the commit history.

:+1: I'll keep that in mind for future Spring Integration contributions.

I've extracted the `RedisLockRegistry` diff applied in the last PR update from my local history:

```patch
--- spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java	(revision )
+++ spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java	(revision )
@@ -17,6 +17,7 @@
 package org.springframework.integration.redis.util;
 
 import java.text.SimpleDateFormat;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.Date;
 import java.util.HashMap;
@@ -31,16 +32,15 @@
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
-import org.springframework.core.io.ClassPathResource;
 import org.springframework.dao.CannotAcquireLockException;
 import org.springframework.data.redis.connection.RedisConnectionFactory;
 import org.springframework.data.redis.core.RedisTemplate;
 import org.springframework.data.redis.core.script.DefaultRedisScript;
 import org.springframework.data.redis.core.script.RedisScript;
 import org.springframework.data.redis.serializer.StringRedisSerializer;
+import org.springframework.integration.support.locks.DefaultLockRegistry;
 import org.springframework.integration.support.locks.ExpirableLockRegistry;
 import org.springframework.integration.support.locks.LockRegistry;
-import org.springframework.scripting.support.ResourceScriptSource;
 import org.springframework.util.Assert;
 
 /**
@@ -78,6 +78,18 @@
 
 	private static final String CLIENT_ID_HASH_KEY = "clientId";
 
+	private static final String OBTAIN_LOCK_SCRIPT =
+			"local lockClientId = redis.call('HGET', KEYS[1], '" + CLIENT_ID_HASH_KEY + "')\n" +
+			"if lockClientId == ARGV[1] then\n" +
+			"  redis.call('PEXPIRE', KEYS[1], ARGV[2])\n" +
+			"  return true\n" +
+			"elseif not lockClientId then\n" +
+			"  redis.call('HMSET', KEYS[1], '" + CLIENT_ID_HASH_KEY + "', ARGV[1])\n" +
+			"  redis.call('PEXPIRE', KEYS[1], ARGV[2])\n" +
+			"  return true\n" +
+			"end\n" +
+			"return false";
+
 	private final Map<String, RedisLock> locks = new HashMap<String, RedisLock>();
 
 	private final String clientId = UUID.randomUUID().toString();
@@ -109,11 +121,25 @@
 		Assert.notNull(connectionFactory, "'connectionFactory' cannot be null");
 		Assert.notNull(registryKey, "'registryKey' cannot be null");
 		this.redisTemplate = createRedisTemplate(connectionFactory);
-		this.obtainLockScript = createObtainLockScript();
+		this.obtainLockScript = new DefaultRedisScript<Boolean>(OBTAIN_LOCK_SCRIPT, Boolean.class);
 		this.registryKey = registryKey;
 		this.expireAfter = expireAfter;
 	}
 
+	/**
+	 * Constructs a lock registry with the supplied lock expiration and a custom local {@link LockRegistry}.
+	 * @param connectionFactory The connection factory.
+	 * @param registryKey The key prefix for locks.
+	 * @param expireAfter The expiration in milliseconds.
+	 * @param localRegistry The local registry used to reduce wait time,
+	 * {@link DefaultLockRegistry} is used by default.
+	 * @deprecated since 4.3.9
+	 */
+	public RedisLockRegistry(RedisConnectionFactory connectionFactory, String registryKey,
+			long expireAfter, LockRegistry localRegistry) {
+		this(connectionFactory, registryKey, expireAfter);
+	}
+
 	private static RedisTemplate<Object, Object> createRedisTemplate(RedisConnectionFactory connectionFactory) {
 		RedisTemplate<Object, Object> template = new RedisTemplate<Object, Object>();
 		template.setConnectionFactory(connectionFactory);
@@ -122,12 +148,18 @@
 		return template;
 	}
 
-	private static RedisScript<Boolean> createObtainLockScript() {
-		DefaultRedisScript<Boolean> script = new DefaultRedisScript<Boolean>();
-		script.setScriptSource(new ResourceScriptSource(
-				new ClassPathResource("org/springframework/integration/redis/util/obtainlock.lua")));
-		script.setResultType(Boolean.class);
-		return script;
+	/**
+	 * Change the state of thread local weak references storage for unlocked locks.
+	 * Thread local weak references are used for lock obtaining optimization -
+	 * thread will get same {@link RedisLock} object for certain key before actual
+	 * locking and after unlocking (if variable still exists).
+	 * <p>While is switched off (by default) every {@link RedisLockRegistry#obtain} call will provide
+	 * different {@link RedisLock} objects for same unlocked key.
+	 * @param useWeakReferences set to true for switch thread local weak references storage on, false by default
+	 * @since 4.0.7
+	 * @deprecated since 4.3.9
+	 */
+	public void setUseWeakReferences(boolean useWeakReferences) {
 	}
 
 	@Override
@@ -160,6 +192,10 @@
 		}
 	}
 
+	public Collection<Lock> listLocks() {
+		return Collections.<Lock>unmodifiableCollection(this.locks.values());
+	}
+
 	private final class RedisLock implements Lock {
 
 		private final String lockKey;
@@ -260,8 +296,8 @@
 
 		private boolean obtainLock() {
 			boolean success = RedisLockRegistry.this.redisTemplate.execute(RedisLockRegistry.this.obtainLockScript,
-					Collections.<Object>singletonList(constructLockKey()), CLIENT_ID_HASH_KEY,
-					RedisLockRegistry.this.clientId, String.valueOf(RedisLockRegistry.this.expireAfter));
+					Collections.<Object>singletonList(constructLockKey()), RedisLockRegistry.this.clientId,
+					String.valueOf(RedisLockRegistry.this.expireAfter));
 			if (success) {
 				this.lockedAt = System.currentTimeMillis();
 			}
```

**Comment by artembilan** [comment]
@kiakimov, hello!

You might be interested in this change. 
At least you did the most hard work for the `RedisLockRegistry`, so maybe you can provide your review as well.
Thank you!

**Comment by kiakimov** [comment]
Hello @artembilan and @vpavic 
I investigated an issue with the sample @vpavic kindly provided, so here is some thoughts:

- RedisLockRegistry does check actual TTL but on obtaining not on actual locking, so from the first sight I would say that it is improper LockRegistry usage - better to obtain the lock in the place it is going to be used. LockRegistryLeaderInitiator obtains it once - in constructor. If we update LockRegistryLeaderInitiator with proper obtaining then sample will work but will produce a warning each time it catches the expire (I checked it).

- From another hand we have a gap between obtaining and actual locking so we have obvious inconsistency - Redis lock can expire right after the obtaining so we will face the same issue @vpavic captured. As for me, I'd rather say that actual Redis lock expire time is some kind of fallback - to be not be blocked on lock holder crash, in regular scenario unlock will be called before expire and locked routine has to be much faster comparing to Redis TTL.

- Purposed solution removes local optimizations and if we do so then current implementation could be simplified to just `SET NX <key> <expire>` (for each `obtain`) and local keys cache. Anyway without Redis events we cant relay on such distributed lock - we still have inconsistency. The only thing we have is polling (LockRegistryLeaderInitiator does it) and post-mortem exception on obtain/unlock. Basically we need to choose what do we want more: local optimizations with some usage tradeoffs (focus on frequent lock usages) or straight Redis pass thru with bigger latencies  (focus on distributed competitions). Anyway both solutions have same inconsistency. As for LockRegistryLeaderInitiator issue - moving `obtain` closer to the `tryLock` and increased TTL will solve the problem without RedisLockRegistry changes - this is the case of distributed competition when local optimizations bring no profit.

P.S. btw looks like LockRegistryLeaderInitiator has one more issue - in the last `finally` around its polling routine it tries to unlock in any case - even if it had no lock - exception is guaranteed.

**Comment by vpavic** [comment]
@artembilan I've updated the tests.

Hi @kiakimov, thanks for sharing.

> RedisLockRegistry does check actual TTL but on obtaining not on actual locking, so from the first sight I would say that it is improper LockRegistry usage - better to obtain the lock in the place it is going to be used.

Can you explain why do you think this isn't a proper usage of `LockRegistry`? I didn't find any javadoc and/or documentation suggesting this.

> I'd rather say that actual Redis lock expire time is some kind of fallback - to be not be blocked on lock holder crash, in regular scenario unlock will be called before expire and locked routine has to be much faster comparing to Redis TTL.

> Purposed solution removes local optimizations and if we do so then current implementation could be simplified to just SET NX <key> <expire> (for each obtain) and local keys cache.

IMO you need to update the lock state in the underlying data store each time it is re-acquired, otherwise the TTL will kick-in to delete the lock which can, and often times will, result in change of leading `LockRegistryLeaderInitiator` instance.

In a system with quite a lot of `SmartLifecycle` components that depend on `LockRegistryLeaderInitiator` to manage them this results in needless stopping and starting of those components, which can be quite costly operation. Once a `LockRegistryLeaderInitiator` instance is granted leadership I wouldn't expect it to lose leadership other than in some sort of exceptional situation (shutdown, crash, etc.).

**Comment by artembilan** [comment]
@kiakimov , 

thank you very much for your valuable feedback!

@vpavic ,

thanks to you too for update and following discussion!

My opinion is:

The `LockRegistry.obtain()` must just return a `Lock` instance, nothing more. Like it is with the `DefaultLockRegistry`, `GemfireLockRegistry`, `ZookeeperLockRegistry`. I don't talk here about `JdbcLockRegistry` because this PR is based on that one and I don't want the "holly war" that the fix conflicts with the previous purpose. 

It's really already the `lock()` responsibility to track expired or not. And I think that really would be better to follow with all other distributed locks solutions: update repository on each `lock()` to actualize the `expire` property. Seems for me that is exactly the purpose of distributed locks without any local optimizations do not involving repository state.

Sorry, @kiakimov , but @vpavic 's fix is more closer to me then the previous (yours) local optimization 😉 

Anyway, would you mind, @kiakimov , to raise a JIRA for that `LockRegistryLeaderInitiator` concern about its `finally` logic? Thanks

**Comment by kiakimov** [comment]
@artembilan mb I was not clear but actually I tried to raise the question do we really need that optimizations and I didnt implement it, I just fixed memory leak there ;)
So if we really dont need this optimization then Im absolutely fine with this decision :)

@vpavic yes, u're right - there is no such suggestions, I use redis locks for a couple of years and I used to obtain it near the actual lock just because of relaying on current implementation, so yes, there is no such requirements

Guys, if you dont mind I'll review the purposed PR in a few days, thank you both for the feedback and discussion

@artembilan for the mentioned issue - https://jira.spring.io/browse/INT-4254

**Comment by artembilan** [comment]
Ah! That was the point, @kiakimov ! 😄 All you do is adapt your application for an existing implementation.
So, that is big "Huh!.." that we all think the same way.

That would be great to have your review, @kiakimov , and maybe some tests against your real applications to be sure before merging. I'd appreciate to have it until this Wednesdays. Otherwise I'm going to vacation since Thurthday and you both will have to explain everything one more time to @garyrussell to enter him into the subject. (If he doesn't follow us silently already, of course 😉 )

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: `- the local {@link LockRegistry} isn't used any more`

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: `- the local optimization isn't used any more because of cluster-wide inconsistency`

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: I think we have to call `constructLockKey()` only once in the ctor. That is really logical that the `RedisLock` instance is tied with the `Registry` and must not be equal to any other instance in different registry even if their keys are the same.
I mean the `equals()` impl here.

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: Why it must be `String`?
Doesn't Redis driver convert it automatically if needed?

File: spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java
Comment (artembilan) [inline spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java]: `registry.expireUnusedOlderThan(-1000);` instead of `Thread.sleep()` 😉 
I don't like any blocking in tests for nothing.

**Review by artembilan — CHANGES_REQUESTED** [review]
That's all from my side.
Excellent work, @vpavic !

Only several minor concerns which I even can address on merge if that.

@kiakimov , I'm sure we will wait for your feedback anyway.

Thank you both!

File: spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java
Comment (vpavic) [inline spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java]: Nice - I guess the same optimization could be done in other places as well (for example `JdbcLockRegistryTests`).

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (vpavic) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: Well, it's important to consider the execution context of Lua script - I've only been able to make things work this way. Revisiting `RedisTemplate` setup, I guess we could switch to `StringRedisTemplate` since that's what we effectively use. It would save a few lines of code as well. WDYT?

**Comment by vpavic** [comment]
@artembilan I've addressed most of your comments, just see [this comment](https://github.com/spring-projects/spring-integration/pull/2114#discussion_r110976706) on use of `RedisTemplate`.

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: 👍 Yes, that sounds reasonable

File: spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java
Comment (artembilan) [inline spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java]: Well, that's true. Although I see only one `Thread.sleep()` there 😄

**Comment by artembilan** [comment]
So far so good, @vpavic .

I have answered to your `StringRedisTemplate` remark.
That is the last thing I could ask you to do.

Only what we need before merging is @kiakimov 's 👍 

I'll do some minor Migration Guide note on the matter. Just in case we are protected from unexpected questions 😄

**Comment by artembilan** [comment]
Here we go: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-4.2-to-4.3-Migration-Guide#redislockregistry !

Doing that I realized that we are missing some important part like lock object in the store.
I see that you removed `LockSerializer`. But what is the alternative now @vpavic ?
Thanks

**Comment by vpavic** [comment]
Well, we don't need a `LockSerializer` since with this new implementation we're dealing with simple `String` values (we're writing a hash using [`HMSET`](https://redis.io/commands/hmset)) hence the `StringRedisSerializer` is now used.

When writing lock to the store, all we need now is lock key, client id and TTL. The first two are natural `String` values while TTL is sent to the wire stringified since [`PEXPIRE`](https://redis.io/commands/pexpire) command that sets it is executed by Lua script handles obtain lock "transaction".

Regarding reading operations, all we actually ever read is client id, again in Lua script and in `RedisLockRegistry.RedisLock#isAcquiredInThisProcess`.

Hope this explains the storage/serialization concerns. This approach in completely analogous to one from `JdbcLockRegistry`.

**Review by kiakimov — COMMENTED** [review]
Hi @artembilan, @vpavic!

Here is my concerns summarizing my review:

1) Instead of GC duty to clear unused locks we move this responsibility to the user, without calling `expireUnusedOlderThan` it is easy to get into troubles (especially if user missed migration guide).

2) We break previous behaviour - it is impossible to get all cluster wide locks with suggested implementation.

3) No guarantee to get same Lock object from the same thread even in sequential calls - coz of `expireUnusedOlderThan` work. It is debatable coz Im not sure is it a big deal or not, but definitely if someone relays on such behaviour he/she wont be happy with new version.

So in scope of raised problem I'd rather suggest leave most of the code and fix just one problem - sudden expire of the remote key protected by local optimization - just fix RedisLock.obtainLock method and do additional Redis read there with skipping entrance local optimization check.

Ofcouse if we would talk about local project my concerns would cost nothing so it is up to you guys to choose the way - backward compatibility or code simplification.

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: 👍 without comments.
We also can consider to address the same in the `JdbcLockRegistry` and `ZookeeperLockRegistry`

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: `ExpirableLockRegistry`:
```
* A {@link LockRegistry} implementing this interface supports the removal of aged locks
* that are not currently locked.
```
Since we can't assume if the lock in the Redis is locked in some other process, we can remove only local instances.
And from the [Reference Manual](http://docs.spring.io/spring-integration/reference/html/zookeeper.html#zk-lock-registry):

> For unbounded strategies (such as the default) you will need to invoke expireUnusedOlderThan(long age) from time to time, to remove old unused locks from memory.

So, we are really talking here only about **in-memory**.

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (artembilan) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: I think this is good catch, @kiakimov !
And it reflects my question about `Lock` serialization, @vpavic .

Although this API is pretty arbitrary and similar in the `JdbcLockRegisrty` returns only local instances, and even any other impls don't provide such an API, I think that should not be so critical to be consistent with existing functionality and make this method to return all the locks from the Redis.

I don't know why why needs it, because there is no use for a `Lock` object obtained in the different process, but that is what we had before.

So, would you mind, @vpavic , to restore `Lock` serialization in some form for this solution? E.g. some simple JSON representation?

Thanks!

**Review by artembilan — CHANGES_REQUESTED** [review]
> without calling expireUnusedOlderThan

Well, we may consider to use `LinkedHashMap.removeEldestEntry()` to protect those people who don't look into migration guide.

>  it is impossible to get all cluster wide locks with suggested implementation.

As I said in the line comment: we should to restore previous behavior in this part

> No guarantee to get same Lock object from the same thread even in sequential calls

Good point! I guess we don't have choice unless document that properly. And I guess the best place is exactly JavaDoc of the `expireUnusedOlderThan()`

> So in scope of raised problem I'd rather suggest leave most of the code and fix just one problem 

That's the good point, too. I think we will come to something closer to the "minimal fix" when all these latest concerns will be addressed.

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (vpavic) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: We should store additional fields in the hash that represents the lock (currently there's only `clientId`), and use [`HGETALL`](https://redis.io/commands/hgetall) to load the entire hash and reconstruct the `RedisLock`.

JSON serialization isn't suitable here as we need to consider the Lua script for obtaining lock - we don't want to parse `clientId` from a JSON in there :smile:

> Although this API is pretty arbitrary and similar in the `JdbcLockRegisrty` returns only local instances, and even any other impls don't provide such an API, I think that should not be so critical to be consistent with existing functionality and make this method to return all the locks from the Redis.

Can you clarify the _similar in the `JdbcLockRegisrty` returns only local instances_ part a bit? AFAICT `JdbcLockRegistry` doesn't offer functionality similar to `RedisLockRegistry#listLocks`.

Also could we consider providing this in `4.3.x`, to maintain backwards compatibility, and deprecate `RedisLockRegistry#listLocks` and remove it in `5.0`? It's not a part of `LockRegistry` contract and as you say _the API is pretty arbitrary_. IMO it's unfortunate that such method forces us to make our storage more complicated than they would've been otherwise.

**Comment by artembilan** [comment]
Well, if understand HMSET properly, we can store several field under that key. So, my point was to leave `clientId` as separate property, but add one more for lock JSON representation, which should be read only and only for that list method. 

On the other hand I really can't imagine the use-case for that list method and how the phantom locks (deserialized instances from othe cluster members) can be useful . More over that can be dangerous because they will be returned with their own remote `clientId`. And what is bad we don't have control over that in the current impl. We might add something like read only RedisLock version if client is different from the Registry deserialing lock .

Any other ideas?

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (vpavic) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: I don't quite agree that `ConcurrentMap#putIfAbsent` does the same since it returns previous value from the map if it puts, rather than the one we've just put into the map. It also forces us to eagerly create `RedisLock` instance meaning with `ConcurrentMap` we have to do something like this:

```java
RedisLock newLock = new RedisLock(path);
RedisLock oldLock = this.locks.putIfAbsent(path, newLock);
return (oldLock == null) ? newLock : oldLock;
```

I'd personally leave it as it is. Thoughts @artembilan?

**Comment by vpavic** [comment]
I've given this some thought over the past 2 days and I didn't like any of the options of implementing this. What also bothers me is that if we implement this as suggested, is that we'd return a different object for the same lock if it was fetched via `RedisLockRegistry#obtain` vs `RedisLockRegistry#listLocks`.

TBH from the first moment `RedisLockRegistry#listLocks` looked like a method for testing purposes that was accidentally made `public` instead of `package-private`. After all, I'd expect users to interact with the lock registry via `LockRegistry`/`ExpirableLockRegistry` and `Lock` APIs.

Like mentioned in one of the [previous comments](https://github.com/spring-projects/spring-integration/pull/2114#discussion_r111316748), I'd prefer to deprecate this in `4.3.x` and remove in `5.0` as we did in two other occasions in this PR.

**Comment by artembilan** [comment]
@vpavic ,

I сompletely share your concerns about `listLocks()` API. 
Yes, we can deprecate it in `4.3.x` and remove in `5.0` as something redundant and confusing.
But! We should have some strong argument do not allow that API.

The `ReentrantLock` is `Serializable`, for example. Therefore can be deserialized in other process and used for its locking purpose. So, the same may be applied for the `RedisLock`, but yeah, we definitely should not let to  use the `RedisLock` created in other client.

I might definitely won't mind to remove that API in `5.0` and simplify the store representation and network communication as you are doing right now, but I insist that we have to provide some compatible fix for the `4.3.x`. 

The  `@Deprecated` ctor and setter do nothing because they don't affect our new internals.
But this `listLocks()` must not change behavior and really return expected list of locks from the store.

That is one vision.

@garyrussell , 

we would like to hear your opinion as well, please.

In two words the problem is like.

When we obtain the lock, each its subsequent `lock()` doesn't update entity in the store and eventually we lose the connection with the `expired` property. The other registry instance decides that the lock is old and locks its local version, too. We see the problem with the `LockRegistryLeaderInitiator`.

The suggested fix is fully similar to the `JdbcLockRegistry` and introduce `clientId` - who is locker.
And also remove some "useless", but introduce breaking changes to the `listLocks()`.

Since the concern is valid, but unfortunately it is so breaking change that the other way to go for `4.3.x` might be really like @kiakimov suggests: 

>  If we update `LockRegistryLeaderInitiator` with proper obtaining then sample will work but will produce a warning each time it catches the expire (I checked it).

All the rest thorough work leave for the `5.0` version.

Thanks

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (vpavic) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: Yes, this seems unnecessary.

File: spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java
Comment (vpavic) [inline spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java]: Now that you mention it I also think we're good with a simple `SET` command. I guess I started with `HMSET` since I expected we'll be storing more attributes and never revisited that.

**Comment by vpavic** [comment]
@artembilan I've addressed your latest concerns.

**Comment by garyrussell** [comment]
I haven't really been following this dicussion, but I am a bit wary of putting such a big change into a point release (4.3.9). I wonder if we should first put it in our upcoming 5.0.0.M4 milestone (in a couple of weeks) and let people kick its tires there and then put it in a 4.3.x release after it's been shaken out.

Or am I being too paranoid?

**Comment by artembilan** [comment]
Let me provide some fix for the `4.3.x` based on the @kiakimov 's suggestion!

Will come back to you soon with an update...

**Comment by artembilan** [comment]
Guys, this is what I suggest as a fix for `4.3.x`: #2121

**Comment by vpavic** [comment]
@artembilan Good to see `4.3.x` has been taken care of with a simpler fix! :+1:

So I guess should I rebase this PR against `master` now? With one additional question - how do we now handle methods that were planned to be deprecate in `4.3.x` and removed in `5.0`?

**Comment by artembilan** [comment]
Sorry, @vpavic , missed that part in my first round. 
Here we go: #2122!

And yes, your branch must be rebased against `master` now and you feel free to go ahead with discussed solution.

**Comment by vpavic** [comment]
@artembilan I've rebased the PR against the `master` branch.

**Review by artembilan — APPROVED** [review]

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/f7f7bdd067428ae570c6bc0cfe489e1698c0a203.

Thank you @vpavic for crazy work!
And especial thanks to @kiakimov for thorough review and ideas for fixes.

Looking forward for more to work with your guys, even if I'm not so useful in the subject 😄

**Pull request closed without merging**
